be/src/exec/pipeline/fragment_executor.cpp

```diff
@@ -302,6 +303,21 @@ Status FragmentExecutor::_prepare_exec_plan(ExecEnv* exec_env, const UnifiedExec
         if (auto* olap_scan = dynamic_cast<vectorized::OlapScanNode*>(scan_node)) {
             olap_scan->enable_shared_scan(enable_shared_scan);
         }
+        if (scan_node->limit() > 0) {
+            sum_scan_limit += scan_node->limit();
+        }
+    }
+
+    if (_wg && _wg->big_query_scan_rows_limit() > 0) {
+        // For SQL like: select * from xxx limit 5, the underlying scan_limit should be 5 * parallelism
+        // Otherwise this SQL would exceed the bigquery_rows_limit due to underlying IO parallelization
+        if (sum_scan_limit <= _wg->big_query_scan_rows_limit()) {
+            int parallelism = dop * ScanOperator::MAX_IO_TASKS_PER_OP;
+            int64_t parallel_scan_limit = sum_scan_limit * parallelism;
+            _query_ctx->set_scan_limit(parallel_scan_limit);
```

Comment:
What if `parallel_scan_limit > _wg->big_query_scan_rows_limit()` ?

Reply:
Then the physical limit would be a larger number.

E.g. The `big_query_scan_rows_limit` is 5, the SQL is `select limit 4`, the `parallel_scan_limit` would be `4 * 8 * 4 = 128`, which is greater than `5`. But the behavior is understandable, cause `limit 4` did not exceed the `big_query_scan_rows_limit`.